# 1. Reading a CSV file using Pandas

In [1]:
from urllib.request import urlretrieve

In [3]:
italy_covid_url = 'https://gist.githubusercontent.com/aakashns/f6a004fa20c84fec53262f9a8bfee775/raw/f309558b1cf5103424cef58e2ecb8704dcd4d74c/italy-covid-daywise.csv'

In [4]:
urlretrieve(italy_covid_url, 'italy-covid-daywise.csv')

('italy-covid-daywise.csv', <http.client.HTTPMessage at 0x190e3180c70>)

In [5]:
import os

In [9]:
os.listdir()

['.ipynb_checkpoints',
 '.jovianrc',
 'data',
 'data2',
 'italy-covid-daywise.csv',
 'movies2.csv',
 'movies3.csv',
 'python-pandas-data-analysis.ipynb',
 'Untitled.ipynb',
 'x.csv']

In [10]:
import pandas as pd

In [11]:
pd.read_csv('italy-covid-daywise.csv')

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

In [12]:
covid_df = pd.read_csv('italy-covid-daywise.csv')

In [13]:
#check the type of covid_df
type(covid_df)

pandas.core.frame.DataFrame

In [14]:
covid_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

## Jyst by looking at the above data what we can tell?


1. counts are provided for each day
2. matrices are new cases, new deaths and new tests for each day
3. above data is of 248 days, starting from 31st dec, 2019 to 3rd sep, 2020

## Q> How to view some basic information about data?
## A> With DataFrame.info() method

In [16]:
# we have a dataframe for the above data, also we can get it with just by reading the csv file using pandas func, pd.read_csv(file)
covid_df = pd.read_csv('italy-covid-daywise.csv')
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        248 non-null    object 
 1   new_cases   248 non-null    float64
 2   new_deaths  248 non-null    float64
 3   new_tests   135 non-null    float64
dtypes: float64(3), object(1)
memory usage: 7.9+ KB


#### Q> What information we get using .info() method on pandas dataframe?
#### A> 
1. Number of rows/entries
2. number of data columns
3. columnwise info of data columns inlcluding 
    * index, 
    * column name,
    * Non-Null Count
    * Dtype
4. memory usage

### Q> How to view statistical information for numerical columns?
### A> with DataFrame.describe() method

### Q> What kind of statistical information we get with .describe() method for numerical columns?
### A> we get
1. count
2. mean
3. standard deviation
4. minimum 
5. 25%, 50%, 75%
6. maximum values

In [17]:
covid_df.describe()

new_cases  new_deaths     new_tests
count   248.000000  248.000000    135.000000
mean   1094.818548  143.133065  31699.674074
std    1554.508002  227.105538  11622.209757
min    -148.000000  -31.000000   7841.000000
25%     123.000000    3.000000  25259.000000
50%     342.000000   17.000000  29545.000000
75%    1371.750000  175.250000  37711.000000
max    6557.000000  971.000000  95273.000000

#### Q> Which property we should use to get the columns within the dataframe?
#### A> dataframe.columns

In [20]:
a = covid_df.columns #returns list of columns

In [21]:
a

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

In [28]:
for i in range(len(a)):
    print(a[i])

date
new_cases
new_deaths
new_tests


#### Q> How to get the number of rows and columns in a dataframe?
#### A> with .shape property we can retrieve number of rows and columns

In [29]:
covid_df.shape

(248, 4)

# 2. Retrieving data from a dataframe

In [30]:
covid_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

#### Retrieve new_cases column

In [31]:
covid_df['new_cases']

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

#### What is the type of retrieved data

In [32]:
type(covid_df['new_cases'])

pandas.core.series.Series

#### How to retrieve specific value

In [33]:
covid_df['new_cases'][246]

975.0

In [35]:
covid_df['new_tests'][111]

7841.0

#### How to retrieve speicifc value using row number and column name 

In [36]:
covid_df.at[111, 'new_tests']

7841.0

In [37]:
covid_df.at[246, 'new_cases']

975.0

#### access column of data frame without bracket notation, with .notation,
#### works only if column name does not contain spaces or special characters

In [38]:
covid_df.new_cases

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

#### If you want multiple columns data at once, pass list of columns in the bracket notation

In [40]:
cases_tests_df = covid_df[['new_cases', 'new_tests']]
cases_tests_df

new_cases  new_tests
0          0.0        NaN
1          0.0        NaN
2          0.0        NaN
3          0.0        NaN
4          0.0        NaN
..         ...        ...
243     1444.0    53541.0
244     1365.0    42583.0
245      996.0    54395.0
246      975.0        NaN
247     1326.0        NaN

[248 rows x 2 columns]

#### Q> How to make full copy of a dataframe?
#### A> with .copy method

In [41]:
covid_copy_df = covid_df.copy()

In [42]:
covid_copy_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

#### Q> covid_df and covid_copy_df are completely separate dataframes
#### A> yes, .copy() method creates completely separate dataframes, changing values inside one of them will not change the other

#### Q> How to access specific row of data
#### A> with .loc method

In [43]:
covid_df.loc[247]

date          2020-09-03
new_cases         1326.0
new_deaths           6.0
new_tests            NaN
Name: 247, dtype: object

In [44]:
covid_df.loc[246]

date          2020-09-02
new_cases          975.0
new_deaths           8.0
new_tests            NaN
Name: 246, dtype: object

#### Q> What is the type of covid_df.loc[index]
#### A> pandas.core.series.Series

In [45]:
#A>
type(covid_df.loc[246])


pandas.core.series.Series

#### Q> Which method allows us to view the few first or few last rows of data in a dataframe?
#### A> .head allows to view first few rows of data and .tail() allows us to view the last few rows

In [46]:
covid_df.head(3)

date  new_cases  new_deaths  new_tests
0  2019-12-31        0.0         0.0        NaN
1  2020-01-01        0.0         0.0        NaN
2  2020-01-02        0.0         0.0        NaN

In [50]:
covid_df.tail(3)

date  new_cases  new_deaths  new_tests
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

In [48]:
type(covid_df.head(3))

pandas.core.frame.DataFrame

#### verify 2nd row has 0 new cases and 1st row has NaN new tests

In [51]:
covid_df.at[1, 'new_cases']

0.0

In [53]:
covid_df.at[0, 'new_tests']

nan

what is the data type of nan values for new tests

In [54]:
type(covid_df.at[0, 'new_tests'])

numpy.float64

#### Q> How to find the first index that doesn't contain NaN value
#### A> using .first_valid_index() method

In [55]:
covid_df.new_tests.first_valid_index()

111

#### verify 111 index has no NaN value for new_tests

In [56]:
covid_df.at[111, 'new_tests']

7841.0

In [57]:
covid_df.at[110, 'new_tests']

nan

#### check the data from 108 to 113 index

In [58]:
covid_df.loc[108:113]

date  new_cases  new_deaths  new_tests
108  2020-04-17     3786.0       525.0        NaN
109  2020-04-18     3493.0       575.0        NaN
110  2020-04-19     3491.0       480.0        NaN
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0

#### Q> How to retrieve random sample of rows from dataframe
### A> with .sample() method, argument is number of samples

In [62]:
covid_df.sample(5)

date  new_cases  new_deaths  new_tests
145  2020-05-24      669.0       119.0    34206.0
99   2020-04-08     3039.0       604.0        NaN
132  2020-05-11      802.0       165.0    25823.0
220  2020-08-07      401.0         6.0    30392.0
140  2020-05-19      451.0        99.0    40226.0

In [66]:
covid_df.loc[99]

date          2020-04-08
new_cases         3039.0
new_deaths         604.0
new_tests            NaN
Name: 99, dtype: object

In [67]:
covid_df.at[99, 'new_tests']

nan